internal/exp/metrics/streams/streams.go

```diff
@@ -0,0 +1,88 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package streams // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/streams"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+)
+
+// Sequence of streams that can be iterated upon
+type Seq[T any] func(yield func(identity.Stream, T) bool) bool
+
+type ErrNotExist struct {
+	Ident identity.Stream
+}
+
+func (e ErrNotExist) Error() string {
+	return fmt.Sprintf("no item for stream %s", e.Ident)
+}
+
+func NotExist(err error) bool {
+	var notExist ErrNotExist
+	return errors.As(err, &notExist)
+}
+
+func Missing(id identity.Stream) error {
+	return ErrNotExist{Ident: id}
+}
+
+func Exists(id identity.Stream, ok bool) error {
+	if !ok {
+		return Missing(id)
+	}
+	return nil
+}
+
+// Map defines a collection of items tracked by a stream-id and the operations
+// on it
+type Map[T any] interface {
+	Load(context.Context, identity.Stream) (T, error)
+	Store(context.Context, identity.Stream, T) error
```

Comment:
While I could see some implementations *could* potentially error during a store operation, the primary players (a raw map, and sync.Map() ) don't. So we're adding a lot of extra error handling for a potential future where someone wants it.